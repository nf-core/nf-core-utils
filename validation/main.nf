#!/usr/bin/env nextflow

// Import plugin functions
include { 
    sayHello;
    getWorkflowVersion;
    checkCondaChannels;
    completionSummary;
    generateVersionReport;
    generateCitationReport;
    processMixedVersionSources;
    processMixedCitationSources;
} from 'plugin/nf-core-utils'

workflow {
    
    log.info """
    ================================
    nf-utils Plugin Integration Test
    ================================
    """
    
    // Test 1: Basic sayHello function
    log.info "Testing sayHello function..."
    sayHello("nf-utils integration test")
    
    // Test 2: Get workflow version
    log.info "Testing getWorkflowVersion function..."
    version = getWorkflowVersion()
    log.info "Workflow version: ${version}"
    
    // Test 3: Check conda channels
    log.info "Testing checkCondaChannels function..."
    conda_check = checkCondaChannels()
    log.info "Conda channels check: ${conda_check}"
    
    // Test 4: Test version processing with sample data
    log.info "Testing version processing..."
    
    // Sample topic channel data: [process, name, version]
    sample_versions = [
        ['FASTQC', 'fastqc', '0.11.9'],
        ['MULTIQC', 'multiqc', '1.12'],
        ['TRIMGALORE', 'trimgalore', '0.6.7'],
        ['TRIMGALORE', 'cutadapt', '4.1']
    ]
    
    // Test version report generation
    version_report = generateVersionReport(sample_versions, [])
    log.info "Version report generated successfully"
    
    // Test 5: Test citation processing with sample data
    log.info "Testing citation processing..."
    
    // Create sample meta.yml files for testing
    sample_meta_paths = []
    
    // Test mixed version sources (topic + legacy)
    log.info "Testing mixed version sources..."
    mixed_versions = processMixedVersionSources(sample_versions, [])
    log.info "Mixed versions processed successfully"
    
    // Test 6: Completion summary
    log.info "Testing completion summary..."
    completionSummary(false)
    
    log.info """
    ================================
    All integration tests completed!
    ================================
    """
}
