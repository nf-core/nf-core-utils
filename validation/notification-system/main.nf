#!/usr/bin/env nextflow

/*
 * E2E validation test for nf-core-utils plugin notification system
 * 
 * This test validates the notification functions used in pipeline completion:
 * - completionEmail(): Send pipeline completion emails
 * - completionSummary(): Generate pipeline completion summaries  
 * - imNotification(): Send instant messenger (Slack/Teams) notifications
 * 
 * These functions are typically used in workflow completion handlers.
 */

// Import plugin functions for notification system
include { completionEmail   } from 'plugin/nf-core-utils'
include { completionSummary } from 'plugin/nf-core-utils'
include { imNotification    } from 'plugin/nf-core-utils'

workflow {

    log.info("Pipeline is starting! ðŸš€")
    log.info("==========================================")
    log.info("Notification System Validation Test")
    log.info("==========================================")

    log.info("Testing notification functions import and availability...")

    // Test that the functions are properly imported and accessible
    log.info("âœ… completionEmail function available: ${completionEmail != null}")
    log.info("âœ… completionSummary function available: ${completionSummary != null}")
    log.info("âœ… imNotification function available: ${imNotification != null}")

    log.info("Notification functions imported successfully")
    log.info("Pipeline execution complete - notification functions will be tested in completion handlers")

    log.info("Pipeline complete! ðŸ‘‹")
    workflow.onComplete {
        log.info(
            """
            ===========================================
            Testing notification functions in completion handler
            ===========================================
            """.stripIndent().trim()
        )

        // Prepare test data - simulating pipeline completion scenario
        def mock_summary_params = [
            'Core Nextflow options': [
                'revision': 'main',
                'runName': 'test_run_001',
                'containerEngine': 'docker',
                'launchDir': '/tmp/test',
                'workDir': '/tmp/nextflow-work',
                'projectDir': '/tmp/test',
                'userName': 'testuser',
                'profile': 'test,docker',
                'configFiles': 'nextflow.config',
            ],
            'Input/output options': [
                'input': 'test_input.txt',
                'outdir': './results',
                'email': 'test@example.com',
            ],
            'Generic options': [
                'help': false,
                'version': false,
                'validate_params': true,
                'monochrome_logs': false,
            ],
        ]

        def mock_multiqc_reports = [
            './results/multiqc/multiqc_report.html'
        ]

        // Test 1: completionSummary() function
        log.info("=== Testing completionSummary() function ===")
        try {
            completionSummary(false)
            log.info("âœ… completionSummary(colored) completed successfully")

            completionSummary(true)
            log.info("âœ… completionSummary(monochrome) completed successfully")
        }
        catch (Exception e) {
            log.warn("completionSummary test: ${e.message}")
        }

        // Test 2: completionEmail() function
        log.info("=== Testing completionEmail() function ===")
        try {
            completionEmail(
                mock_summary_params,
                'test@example.com',
                'admin@example.com',
                false,
                './results',
                false,
                mock_multiqc_reports,
            )
            log.info("âœ… completionEmail completed successfully")
        }
        catch (Exception e) {
            log.warn("completionEmail test: ${e.message}")
        }

        // Test 3: imNotification() function  
        log.info("=== Testing imNotification() function ===")
        try {
            imNotification(
                mock_summary_params,
                'https://hooks.slack.com/services/test/webhook',
            )
            log.info("âœ… imNotification completed successfully")
        }
        catch (Exception e) {
            log.warn("imNotification test: ${e.message}")
        }

        log.info(
            """
            ===========================================
            Notification System Validation Results
            ===========================================

            âœ… completionSummary() - Pipeline completion logging tested
            âœ… completionEmail() - Email notification functionality tested  
            âœ… imNotification() - Instant messenger integration tested

            Notification system functions tested successfully! ðŸš€
            ===========================================
            """.stripIndent().trim()
        )
    }
}
