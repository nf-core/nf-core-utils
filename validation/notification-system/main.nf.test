nextflow_pipeline {

    name "Test nf-core-utils notification system"
    script "./main.nf"

    test("Notification system") {

        then {
            assert workflow.success
            assert snapshot(
                filterNextflowOutput(workflow.stdout + workflow.stderr, keepAnsi: true)
            ).match()
        }
    }
}
