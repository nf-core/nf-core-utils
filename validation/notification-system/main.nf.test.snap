{
    "notification_system_success": {
        "content": [
            [
                "Pipeline is starting! \ud83d\ude80",
                "Pipeline complete! \ud83d\udc4b"
            ],
            0,
            true
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-08-22T14:01:44.640419"
    },
    "notification_testing_complete": {
        "content": [
            true,
            0
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-08-22T13:52:36.215781"
    },
    "notification_validation_success": {
        "content": [
            true,
            0
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-08-22T13:52:38.444984"
    }
}